syntax = "proto3";
import "google/protobuf/wrappers.proto";  
package exchange;

// The gRPC service definition
service ExchangeService {
  rpc SendMessage (MessageRequest) returns (Acknowledge);
  rpc SendResponse (MessageResponse) returns (Acknowledge);
  rpc RegisterClient (ClientRegistrationRequest) returns (stream MessageRequest);
 
}


message RequestId {
  string transaction_id = 1;
}

message MessageRequest {
   google.protobuf.StringValue transaction_id = 1; //nullable filed, Keep null in case of broadcaset message
  string message_id = 2;
  string sender = 3;
   google.protobuf.StringValue receiver = 4;//nullable filed, Keep null in case of broadcaset message
  string message_payload = 5;
}

message ClientRegistrationRequest {
  string client_id = 1;
}

message Acknowledge {
  bool success = 1;
  string reply = 2;
}


message MessageResponse {
  string transaction_id = 1;
  bool result = 2;
  string data = 3;
}